### PR TITLE
fix(types): fetch method should not return void but undefined instead

### PR DESCRIPTION
**Pull request by Thenkei**
Linked to the issue https://github.com/isaacs/node-lru-cache/issues/302

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1919,7 +1919,7 @@ export class LRUCache<K extends {}, V extends {}, FC = unknown> {
     }
 
     const pcall = (
```
Comment (Thenkei) [inline src/index.ts:1921]: My fix is pretty bad.

I think we should use DerredPromise here to have a cleaner sense of what we except.

```
class Deferred<T> {
  promise: Promise<T>;
  resolve: (result: T) => void;
  reject: (error: Error) => void;

  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.reject = reject;
      this.resolve = resolve;
    });
  }
}
```

File: src/index.ts [hunk 2]
```diff
@@ -1934,10 +1934,11 @@ export class LRUCache<K extends {}, V extends {}, FC = unknown> {
...
           options.allowStaleOnFetchAbort
         ) {
-          res()
-          // when it eventually resolves, update the cache.
+          // when it eventually resolves, update the cache or simply resolve undefined.
           if (options.allowStaleOnFetchAbort) {
             res = v => cb(v, true)
+          } else {
+            res(undefined)
           }
```
Comment (Thenkei) [inline src/index.ts:1942]: Was it the expected behavior to resolve `void` and then resolve the value? How could it eventually consume the second resolve?

Comment (Thenkei) [reply src/index.ts]: Reverted as it breaks all tests

**Comment by Thenkei** [comment]
The test suite is now running correctly locally.

**Comment by isaacs** [comment]
[This patch](https://gist.github.com/isaacs/dac8bd3b9afe4c42d5e549afa2ecc16e) would allow the `fetchMethod` to still return `void`, but I don't think there's any way around the SemVer major `LRUCache.fetch` type signature change.

**Comment by Thenkei** [comment]
Thank you so much @isaacs. 🙏 

I hope it will help other TS users too.

**Pull request closed without merging**